modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageWriteHandler.java

```diff
@@ -310,6 +319,28 @@ boolean beforeApply(Command command) {
         return false;
     }
 
+    /**
+     * The callback that is called right after storage is updated with a snapshot.
+     */
+    void onSnapshotLoad() {
+        byte[] keyFrom = IDEMPOTENT_COMMAND_PREFIX_BYTES;
+        byte[] keyTo = storage.nextKey(IDEMPOTENT_COMMAND_PREFIX_BYTES);
+
+        Cursor<Entry> cursor = storage.range(keyFrom, keyTo);
+        // It's fine to loose original command start time - in that case we will store the entry a little bit longer that necessary.
+        HybridTimestamp now = clusterTime.now();
+
+        try (cursor) {
+            for (Entry entry : cursor) {
+                if (!entry.tombstone()) {
+                    byte[] commandIdBytes = copyOfRange(entry.key(), IDEMPOTENT_COMMAND_PREFIX_BYTES.length, entry.key().length);
+                    CommandId commandId = ByteUtils.fromBytes(commandIdBytes);
+                    idempotentCommandCache.put(commandId, new IdempotentCommandCachedResult(byteToBoolean(entry.value()[0]), now));
```

Comment:
Let's add an assertion that the length of `entry.value()` is always 1.

Reply:
It's possible to have both boolean and StatementResult there. Reworked the code.